File: python/rules/dmn.py
from event import Event


class HisVolume:
    def __init__(self):
        self.d_sum = 0
        self.d_curVol = 0
        self.d_hisVol = list()

    def increment(self, vol):
        self.d_curVol += vol

    def update(self):
        if self.d_curVol == 0:
            return False
        self.d_sum += self.d_curVol
        self.d_hisVol.append(self.d_curVol)
        self.d_curVol = 0
        if len(self.d_hisVol) == 5:
            self.d_sum -= self.d_hisVol[0]
            self.d_hisVol.pop(0)
        return True;

    def getSumAndCnt(self):
        return self.d_sum, len(self.d_hisVol)

        
        
class DMN:
    def __init__(self):
        self.d_curVolDict = dict()
        self.d_curDate = -1
        self.d_hisVolDict = dict()

    def updateHis(self, curDate):
        if curDate == self.d_curDate:
            return
        self.d_curDate = curDate
        self.d_curVolDict.clear()
        for key,val in self.d_hisVolDict.items():
            if val.update():
                self.d_hisVolDict[key] = val
        
    def check(self, evn):
        if evn.eventType != "EXECUTE":
            return None

        curDate = evn.dateTime.date()
        self.updateHis(curDate)
        key = evn.broker + "|" + evn.parsekey
        if key not in self.d_curVolDict:
            self.d_curVolDict[key] = 0
        self.d_curVolDict[key] += evn.volume
        tVol = self.d_curVolDict[key] 
        if evn.parsekey not in self.d_hisVolDict:
            self.d_hisVolDict[evn.parsekey] = HisVolume()
        self.d_hisVolDict[evn.parsekey].increment(evn.volume)
        tsum, cnt = self.d_hisVolDict[evn.parsekey].getSumAndCnt()
        if cnt>0 and cnt * tVol > 1.5 * tsum:
            ss = "D M N: {} has executed over 50% of the average volume for '{}', the total volume {}, the average is {:.2f}"
            ss = ss.format(evn.broker, evn.parsekey, tVol, tsum/cnt)
            return ss
            
        return None

File: golang/src/rules/dmn.go
package rule

import (
    ev "events"
    "fmt"
    "time"
)

type hisVolList struct{
    totalVolume int
    curVolume   int
    qhisVolume  []int
}
func makeHisVolList() *hisVolList{
    his := hisVolList{}
    his.totalVolume = 0
    his.curVolume = 0
    his.qhisVolume = make([]int, 0, 4)
    return &his
}
func (his *hisVolList)increment(vol int){
    his.curVolume += vol
}
func (his *hisVolList)update(){
    if his.curVolume == 0 {
        return
    }
    if len(his.qhisVolume) == 4{
        his.totalVolume -= his.qhisVolume[0]
        his.qhisVolume = his.qhisVolume[1:]
    }
    his.totalVolume += his.curVolume
    his.qhisVolume = append(his.qhisVolume,his.curVolume)
    his.curVolume = 0
}
func (his *hisVolList)getAverage()(tVol, cnt int){
    cnt = len(his.qhisVolume)
    tVol = his.totalVolume
    return
}
type Rdmn struct{
    dailyVolume map[string]*int
    hisVolume map[string]*hisVolList
    curY,curD int
    curM time.Month
}
func getDailyKey(evn *ev.Event)string{
    return evn.Broker + "|" + evn.Security
}
func (r *Rdmn)Rupdate(nd * time.Time){
    y,m,d := nd.Date()
    if r.curD == d && r.curM == m && r.curY == y {
        return
    }
    r.curY = y
    r.curM = m
    r.curD = d
    r.dailyVolume = make(map[string]*int)
    for _,v := range r.hisVolume {
        v.update()
    }
}
func InitRdmn()Rule{
    rdmn := Rdmn{}
    rdmn.dailyVolume = make(map[string]*int)
    rdmn.hisVolume = make(map[string]*hisVolList)
    return &rdmn
}

func (r *Rdmn) CheckEvent(evn *ev.Event) (sr string){
    if evn.EventType != ev.EVENT_EXECUTE {
        return
    }
    r.Rupdate(&evn.EventTime)
    nVol := evn.Volume
    key := getDailyKey(evn)
    daVol,ok := r.dailyVolume[key]
    if !ok {
        r.dailyVolume[key] = &nVol
        daVol = &nVol
    } else {
        *daVol += nVol
    }
    hisVol,ok := r.hisVolume[evn.Security]
    if !ok {
        hisVol = makeHisVolList()
        r.hisVolume[evn.Security] = hisVol
    }
    hisVol.increment(evn.Volume)
    tlVol, cnt := hisVol.getAverage()
    if cnt > 0 && float64(*daVol * cnt) > float64(1.5 * float64(tlVol)) {
        sr = fmt.Sprintf("D M N: %v has executed over 50%% of the average volume for '%v', the total volume %v, the average is %.2f",
             evn.Broker, evn.Security, *daVol, (float64(tlVol))/(float64(cnt)))
    }
    return
}

